### ISSUE TITLE
Why model can't converge when batch size > 1?

### ISSUE THREAD

**Issue by Runist** [tensorflow/tensorflow#45838]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, but only demo.
- OS Platform and Distribution (e.g., Linux Ubuntu 18.04):
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): tf-nightly-2.5.0
- Python version: 3.7.4
- CUDA/cuDNN version: CUDA 11.1 / cuDNN 8.0.4
- GPU model and memory: RTX3090 24GB
- RAM: 32GB

This question seems lkie not related to the environment.This is behaviour:

I wrote a code to build the Faster RCNN model. I train it by set batch size = 1. Model can converge, and predict exactly.
But when I set batch size = 2 or greater, The faster_rcnn_regr_loss(smooth L1)  can't converge.

At first I thought it was my something wrong with the program. However I checked my code, I didn't find any error. So I try to  set batch size = 8 but still backpropagation one sample losses in one training step, like this:

**Batch size = 8, backpropagation 8 sample average loss in one training step**
**Mode can not converge**
```
loss_class = classifier_train(model_classifier,
                              [image[valid_roi], x_roi],
                              [y_class_label, y_classifier])
```

**Batch size = 8, backpropagation 1 sample losses in one training step**
**Mode can converge**
```
loss_class = [0, 0]
for j in range(len(valid_roi)):
    loss = classifier_train(model_classifier,
                            [np.expand_dims(image[valid_roi[j]], axis=0),
                             np.expand_dims(x_roi[j], axis=0)],
                            [np.expand_dims(y_class_label[j], axis=0),
                             np.expand_dims(y_classifier[j], axis=0)])
    loss_class[0] += loss[0]
    loss_class[1] += loss[1]

loss_class[0] /= len(valid_roi)
loss_class[1] /= len(valid_roi)
```

**Batch size = 8, backpropagation 1 sample losses 1 times in one training step**
**Mode can converge**
```
loss_class = classifier_train(model_classifier,
                              [np.expand_dims(image[valid_roi[0]], axis=0),
                               np.expand_dims(x_roi[0], axis=0)],
                              [np.expand_dims(y_class_label[0], axis=0),
                               np.expand_dims(y_classifier[0], axis=0)])
```
I think this is a bug.

This is my faster_rcnn_regr_loss fuction:
```
def class_loss_regr(num_classes):
    epsilon = 1e-4

    def class_loss_regr_fixed_num(y_true, y_pred):
        """

        :param y_true:  [batch_size, num_rois, num_classes * 8]
                        [:, :, :num_classes * 4] is label index, [:, :, num_classes * 4:] is ground true boxes coordinate.
        :param y_pred: [batch_size, num_rois, num_classes * 4]
        :return: classifier regr_loss
        """
        regr_loss = 0
        batch_size = len(y_true)
        for i in range(batch_size):
            x = y_true[i, :, 4 * num_classes:] - y_pred[i, :, :]                  
            x_abs = backend.abs(x)                                                  
            x_bool = backend.cast(backend.less_equal(x_abs, 1.0), 'float32')       

            loss = 4 * backend.sum(
                y_true[i, :, :4 * num_classes] * (x_bool * (0.5 * x * x) + (1 - x_bool) * (x_abs - 0.5))) / backend.sum(
                epsilon + y_true[i, :, :4 * num_classes])
            regr_loss += loss

        return regr_loss / backend.constant(batch_size)

    return class_loss_regr_fixed_num
```

This is my training code:
```
def main():
    global rpn_optimizer, classifier_optimizer
    os.environ["CUDA_VISIBLE_DEVICES"] = "0"

    gpus = tf.config.experimental.list_physical_devices("GPU")
    if gpus:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)

    img_input = Input(shape=(None, None, 3))
    roi_input = Input(shape=(None, 4))

    share_layer = ResNet50(img_input)
    rpn = frcnn.rpn(share_layer, num_anchors=len(cfg.anchor_box_ratios) * len(cfg.anchor_box_scales))
    classifier = frcnn.classifier(share_layer, roi_input, cfg.num_rois, nb_classes=cfg.num_classes)

    model_rpn = models.Model(img_input, rpn)
    model_classifier = models.Model([img_input, roi_input], classifier)
    model_all = models.Model([img_input, roi_input], rpn + classifier)

    anchors = get_anchors(cfg.share_layer_shape, cfg.input_shape)

    box_parse = BoundingBox(anchors, max_threshold=cfg.rpn_max_overlap, min_threshold=cfg.rpn_min_overlap)

    reader = DataReader(cfg.annotation_path, box_parse, cfg.batch_size)
    train_data = reader.generate()
    train_step = len(reader.train_lines) // cfg.batch_size

    losses = np.zeros((train_step, 4))
    best_loss = np.Inf

    rpn_lr = CosineAnnealSchedule(cfg.epoch, train_step, cfg.rpn_lr_max, cfg.rpn_lr_min)
    cls_lr = CosineAnnealSchedule(cfg.epoch, train_step, cfg.cls_lr_max, cfg.cls_lr_min)

    rpn_optimizer = optimizers.Adam(rpn_lr)
    classifier_optimizer = optimizers.Adam(cls_lr)

    for e in range(cfg.epoch):
        invalid_data = 0        
        print("Learning rate adjustment, rpn_lr: {}, cls_lr: {}".
              format(rpn_optimizer._decayed_lr("float32").numpy(),
                     classifier_optimizer._decayed_lr("float32").numpy()))


        progbar = utils.Progbar(train_step)
        print('Epoch {}/{}'.format(e+1, cfg.epoch))
        for i in range(train_step):

            image, rpn_y, bbox = next(train_data)
            loss_rpn = rpn_train(model_rpn, image, rpn_y)
            predict_rpn = model_rpn(image)

            predict_boxes = box_parse.detection_out(predict_rpn, confidence_threshold=0)
            height, width = np.shape(image[0])[:2]
            x_roi, y_class_label, y_classifier, valid_roi = get_classifier_train_data(predict_boxes,
                                                                                      bbox,
                                                                                      width,
                                                                                      height,
                                                                                      cfg.batch_size,
                                                                                      cfg.num_classes)

            invalid_data += (cfg.batch_size - len(valid_roi))
            if len(x_roi) == 0:
                progbar.update(i+1, [('rpn_cls', np.mean(losses[:i+1, 0])),
                                     ('rpn_regr', np.mean(losses[:i+1, 1])),
                                     ('detector_cls', np.mean(losses[:i+1, 2])),
                                     ('detector_regr', np.mean(losses[:i+1, 3]))])
                continue
            
            #######you can test code here#########
            loss_class = classifier_train(model_classifier,
                                          [image[valid_roi], x_roi],
                                          [y_class_label, y_classifier])
            #################################

            losses[i, 0] = loss_rpn[0].numpy()
            losses[i, 1] = loss_rpn[1].numpy()
            losses[i, 2] = loss_class[0].numpy()
            losses[i, 3] = loss_class[1].numpy()

            progbar.update(i+1, [('rpn_cls', np.mean(losses[:i+1, 0])),
                                 ('rpn_regr', np.mean(losses[:i+1, 1])),
                                 ('detector_cls', np.mean(losses[:i+1, 2])),
                                 ('detector_regr', np.mean(losses[:i+1, 3]))])

        else:
            loss_rpn_cls = np.mean(losses[:, 0])
            loss_rpn_regr = np.mean(losses[:, 1])
            loss_class_cls = np.mean(losses[:, 2])
            loss_class_regr = np.mean(losses[:, 3])

            curr_loss = loss_rpn_cls + loss_rpn_regr + loss_class_cls + loss_class_regr

            print('\nLoss RPN classifier: {:.4f}'.format(loss_rpn_cls))
            print('Loss RPN regression: {:.4f}'.format(loss_rpn_regr))
            print('Loss Detector classifier: {:.4f}'.format(loss_class_cls))
            print('Loss Detector regression: {:.4f}'.format(loss_class_regr))
            print("{} picture can't detect any roi.".format(invalid_data))

            print('The best loss is {:.4f}. The current loss is {:.4f}.'.format(best_loss, curr_loss))
            if curr_loss < best_loss:
                best_loss = curr_loss

            print('Saving weights.\n')
            model_all.save_weights("./logs/model/voc_{:.4f}.h5".format(curr_loss))
```
I try to show it colab,but I can't download the VOC dataset.And it only occured in my project.I can't write demo to show it because faster_rcnn_regr need rpn predict data.
So maybe you should download my [code](https://github.com/Runist/Faster_RCNN).And you should do this:

1. Download the code.
2. Run this code to get voc annotation 'train.txt' file:
```
iimport xml.etree.ElementTree as ET
import os

def convert_annotation(image_id, list_file, class_names):
    in_file = open('/content/VOCdevkit/VOC2012/Annotations/%s.xml' % (image_id))
    tree = ET.parse(in_file)
    root = tree.getroot()

    for obj in root.iter('object'):

        difficult = obj.find('difficult').text
        cls = obj.find('name').text

        cls_id = class_names.index(cls)
        xmlbox = obj.find('bndbox')

        b = (int(xmlbox.find('xmin').text),
            int(xmlbox.find('ymin').text),
            int(xmlbox.find('xmax').text),
            int(xmlbox.find('ymax').text))

        list_file.write(" " + ",".join([str(a) for a in b]) + ',' + str(cls_id))


if __name__ == '__main__':
    label = ['aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair', 'cow', 'diningtable', 'dog',
         'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa', 'train', 'tvmonitor']
    
    xmlfilepath = '/content/VOCdevkit/VOC2012/Annotations'
    temp_xml = os.listdir(xmlfilepath)
    total_xml = []

    for xml in temp_xml:
        if xml.endswith(".xml"):
            total_xml.append(xml[:-4])

    files = open('train.txt', 'w')
    for image_id in total_xml:
        files.write('/content/VOCdevkit/VOC2012/JPEGImages/{}.jpg'.format(image_id))
        convert_annotation(image_id, files, label)
        files.write('\n')
    files.close()
```
3. Move the train.txt to './config/' directory.
4. Edit the './config/config.py' some training paramters.
5. Run 'python train.py'.

**Comment by gowthamkpr** [comment]
@Runist Can you try it with the latest version of tensorflow (2.8.0) and check if you are facing the same issue. Also please follow this [tutorial](https://gilberttanner.com/blog/tensorflow-object-detection-with-tensorflow-2-creating-a-custom-model) and check if there are any significant differences in the code. Thanks!
